massa-execution-worker/src/execution.rs

```diff
@@ -1542,65 +1559,101 @@ impl ExecutionState {
     }
 
     /// Get every final and active datastore key of the given address
+    #[allow(clippy::type_complexity)]
     pub fn get_final_and_candidate_datastore_keys(
         &self,
         addr: &Address,
-    ) -> (BTreeSet<Vec<u8>>, BTreeSet<Vec<u8>>) {
+        prefix: &[u8],
+    ) -> (Option<BTreeSet<Vec<u8>>>, Option<BTreeSet<Vec<u8>>>) {
         // here, get the final keys from the final ledger, and make a copy of it for the candidate list
         // let final_keys = final_state.read().ledger.get_datastore_keys(addr);
         let final_keys = self
             .final_state
             .read()
             .ledger
-            .get_datastore_keys(addr)
-            .unwrap_or_default();
+            .get_datastore_keys(addr, prefix);
+
         let mut candidate_keys = final_keys.clone();
 
-        // here, traverse the history from oldest to newest, applying additions and deletions
+        // compute prefix range
+        let prefix_range = if !prefix.is_empty() {
+            // compute end of prefix range
+            let n_keep = prefix
+                .iter()
+                .enumerate()
+                .rev()
+                .find_map(|(i, v)| if v < &255 { Some(i + 1) } else { None })
+                .unwrap_or(0);
+            let mut prefix_end = prefix[..n_keep].to_vec();
+            if let Some(v) = prefix_end.last_mut() {
```

Comment:
The proposed alternative from #4170 was incomplete because it states that `[]` is the end for the prefix `[255, 255, 255]`. 

I unified all the behaviors in this PR.